utilslib/dr.py

```diff
@@ -380,18 +391,80 @@ def get_custom_resources(self):
 
     @lib.timing_wrapper
     def save_namespace(self, namespace):
-        count_stored = 0
+        """Save a namespace to S3
+
+        This method will enumerate all resources in a namespace
+        and then backup the yaml to S3. It will also delete
+        any yaml in S3 for resources that no longer exist in
+        the namespace.
+        
+        Arguments:
+            namespace {str} -- the kubernetes namespace to backup
+        
+        Returns:
+            [int] -- number of resources backuped to S3
+            [int] -- number of resources deleted from s3
+        """
+        if not isinstance(namespace, str):
+            raise Exception("namespace must be a string")
+
+        if namespace == "":
+            raise Exception("you must supply a namespace, empty string supplied")
+
+        keys_stored = self._save_to_s3(namespace)
+        keys_deleted = self._handle_deleted_resources(keys_stored, namespace)
+
+        return len(keys_stored), len(keys_deleted)
+
+    @lib.timing_wrapper
+    def _save_to_s3(self, namespace):
+        """Save Kubernetes resources for a namespace to S3
+        
+        Arguments:
+            namespace {str} -- the kubernetes namespace to backup
+        
+        Returns:
+            [str[]] -- an array of keys for the objects stored
+        """
+        keys = []
+        
+        # Save the namespace first
         ns = self.k8s.read_namespace(namespace)
         key, data = self.create_key_yaml(ns)
         self.store.store_in_bucket(key, data)
-        count_stored += 1
-        for kind in K8s.kinds.keys():
+        keys.append(key)
+
+        # Save the kinds that are supported
+        for kind in K8s.supported_kinds.keys():
             for item in self.k8s.list_kind(namespace, kind):
                 read_data = self.k8s.read_kind(namespace, kind, item.metadata.name)
                 key, data = self.create_key_yaml(read_data)
                 self.store.store_in_bucket(key, data)
-                count_stored += 1
-        return count_stored
+                keys.append(key)
+        return keys
+
+    @lib.timing_wrapper
+    def _handle_deleted_resources(self, existing_keys, namespace):
```

Comment:
it might be safer if this function did a k8s api read to verify the existence of the item rather than relying on the backup list. The admission controller updater might have added something since. There is always going to be a window but best make it as small as possible?

Reply:
I'm not sure thats needed. Its just done a k8s read of the namespace and the list of keys comes from that. There is always going to be a change that its not consistent. I'll add an issue though to discuss.